pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AssignmentInOperandRule.java

```diff
@@ -85,31 +85,31 @@ public Object visit(ASTUnaryExpression node, Object data) {
     private void checkAssignment(ASTExpression impureExpr, RuleContext ctx) {
         ASTExpression toplevel = JavaAstUtils.getTopLevelExpr(impureExpr);
         JavaNode parent = toplevel.getParent();
-        if (parent instanceof ASTExpressionStatement) {
+
+        if (toplevel == impureExpr
+            && (parent instanceof ASTExpressionStatement
+            || parent instanceof ASTStatementExpressionList)
+        ) {
             // that's ok
             return;
         }
-        if (parent instanceof ASTIfStatement && !getProperty(ALLOW_IF_DESCRIPTOR)
-            || parent instanceof ASTWhileStatement && !getProperty(ALLOW_WHILE_DESCRIPTOR)
-            || parent instanceof ASTDoStatement && !getProperty(ALLOW_DO_WHILE_DESCRIPTOR)
-            || parent instanceof ASTSwitchStatement && !getProperty(ALLOW_SWITCH_DESCRIPTOR)
-            || parent instanceof ASTForStatement && ((ASTForStatement) parent).getCondition() == toplevel && !getProperty(ALLOW_FOR_DESCRIPTOR)) {
-
-            ctx.addViolation(impureExpr);
+        if (parent instanceof ASTIfStatement && getProperty(ALLOW_IF_DESCRIPTOR)
+            || parent instanceof ASTWhileStatement && getProperty(ALLOW_WHILE_DESCRIPTOR)
+            || parent instanceof ASTDoStatement && getProperty(ALLOW_DO_WHILE_DESCRIPTOR)
+            || parent instanceof ASTSwitchStatement && getProperty(ALLOW_SWITCH_DESCRIPTOR)
+            || parent instanceof ASTForStatement && ((ASTForStatement) parent).getCondition() == toplevel && getProperty(ALLOW_FOR_DESCRIPTOR)
+        ) {
+            return;
         }
+        // report everything else
+        ctx.addViolation(impureExpr);
     }
 
+    @Deprecated
```

Comment:
```suggestion
    /**
     * @deprecated Since 7.17.0. This method is an implementation detail and will be internalized.
     */
    @Deprecated
```